# **Airline Customer Satisfaction Capstone**

## Data Wrangling

In this project, I will work with the data set from https://www.kaggle.com/datasets/mysarahmadbhat/airline-passenger-satisfaction.

The project is split into four separate sections:
- Data wranging
- Exploratory data analysis
- Preprocessing and training
- Modeling

As this is the first section, data wrangling, I will focus on importing the data, cleaning/wrangling the data, and saving the data.

## 1. Import Packages

In [46]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os

## 2. Import Data

In [47]:
airline_data = pd.read_csv('/Users/lauren/Desktop/airline_passenger_satisfaction.csv')

In [48]:
#Call the info method on airline_data to see a summary of the data
airline_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129880 entries, 0 to 129879
Data columns (total 24 columns):
 #   Column                                  Non-Null Count   Dtype  
---  ------                                  --------------   -----  
 0   ID                                      129880 non-null  int64  
 1   Gender                                  129880 non-null  object 
 2   Age                                     129880 non-null  int64  
 3   Customer Type                           129880 non-null  object 
 4   Type of Travel                          129880 non-null  object 
 5   Class                                   129880 non-null  object 
 6   Flight Distance                         129880 non-null  int64  
 7   Departure Delay                         129880 non-null  int64  
 8   Arrival Delay                           129487 non-null  float64
 9   Departure and Arrival Time Convenience  129880 non-null  int64  
 10  Ease of Online Booking                  1298

## 2.1 Explore the Data

In [49]:
#Call the head method on airline_data to print the first several rows of the data
airline_data.head()

ID  Gender  Age Customer Type Type of Travel     Class  Flight Distance  \
0   1    Male   48    First-time       Business  Business              821   
1   2  Female   35     Returning       Business  Business              821   
2   3    Male   41     Returning       Business  Business              853   
3   4    Male   50     Returning       Business  Business             1905   
4   5  Female   49     Returning       Business  Business             3470   

   Departure Delay  Arrival Delay  Departure and Arrival Time Convenience  \
0                2            5.0                                       3   
1               26           39.0                                       2   
2                0            0.0                                       4   
3                0            0.0                                       2   
4                0            1.0                                       3   

   ...  On-board Service  Seat Comfort  Leg Room Service  Cleanliness  \
0  ...                 3             5                 2            5   
1  ...                 5             4                 5            5   
2  ...                 3             5                 3            5   
3  ...                 5             5                 5            4   
4  ...                 3             4                 4            5   

   Food and Drink  In-flight Service  In-flight Wifi Service  \
0               5                  5                       3   
1               3                  5                       2   
2               5                  3                       4   
3               4                  5                       2   
4               4                  3                       3   

   In-flight Entertainment  Baggage Handling             Satisfaction  
0                        5                 5  Neutral or Dissatisfied  
1                        5                 5                Satisfied  
2                        3                 3                Satisfied  
3                        5                 5                Satisfied  
4                        3                 3                Satisfied  

[5 rows x 24 columns]

In [50]:
#Filter the airline_data dataframe to display just the row for ID 1

airline_data[airline_data.ID == 1].T

0
ID                                                            1
Gender                                                     Male
Age                                                          48
Customer Type                                        First-time
Type of Travel                                         Business
Class                                                  Business
Flight Distance                                             821
Departure Delay                                               2
Arrival Delay                                               5.0
Departure and Arrival Time Convenience                        3
Ease of Online Booking                                        3
Check-in Service                                              4
Online Boarding                                               3
Gate Location                                                 3
On-board Service                                              3
Seat Comfort                                                  5
Leg Room Service                                              2
Cleanliness                                                   5
Food and Drink                                                5
In-flight Service                                             5
In-flight Wifi Service                                        3
In-flight Entertainment                                       5
Baggage Handling                                              5
Satisfaction                            Neutral or Dissatisfied

In [51]:
#Filter the airline_data dataframe to display just the row for ID 2

airline_data[airline_data.ID == 2].T

1
ID                                              2
Gender                                     Female
Age                                            35
Customer Type                           Returning
Type of Travel                           Business
Class                                    Business
Flight Distance                               821
Departure Delay                                26
Arrival Delay                                39.0
Departure and Arrival Time Convenience          2
Ease of Online Booking                          2
Check-in Service                                3
Online Boarding                                 5
Gate Location                                   2
On-board Service                                5
Seat Comfort                                    4
Leg Room Service                                5
Cleanliness                                     5
Food and Drink                                  3
In-flight Service                               5
In-flight Wifi Service                          2
In-flight Entertainment                         5
Baggage Handling                                5
Satisfaction                            Satisfied

In [52]:
#Filter the airline_data dataframe to display just the row for ID 3

airline_data[airline_data.ID == 3].T

2
ID                                              3
Gender                                       Male
Age                                            41
Customer Type                           Returning
Type of Travel                           Business
Class                                    Business
Flight Distance                               853
Departure Delay                                 0
Arrival Delay                                 0.0
Departure and Arrival Time Convenience          4
Ease of Online Booking                          4
Check-in Service                                4
Online Boarding                                 5
Gate Location                                   4
On-board Service                                3
Seat Comfort                                    5
Leg Room Service                                3
Cleanliness                                     5
Food and Drink                                  5
In-flight Service                               3
In-flight Wifi Service                          4
In-flight Entertainment                         3
Baggage Handling                                3
Satisfaction                            Satisfied

In [54]:
#Filter the airline_data dataframe to display just the row for ID 4

airline_data[airline_data.ID == 4].T

3
ID                                              4
Gender                                       Male
Age                                            50
Customer Type                           Returning
Type of Travel                           Business
Class                                    Business
Flight Distance                              1905
Departure Delay                                 0
Arrival Delay                                 0.0
Departure and Arrival Time Convenience          2
Ease of Online Booking                          2
Check-in Service                                3
Online Boarding                                 4
Gate Location                                   2
On-board Service                                5
Seat Comfort                                    5
Leg Room Service                                5
Cleanliness                                     4
Food and Drink                                  4
In-flight Service                               5
In-flight Wifi Service                          2
In-flight Entertainment                         5
Baggage Handling                                5
Satisfaction                            Satisfied

In [10]:
#Count (using `.sum()`) the number of missing values (`.isnull()`) in each column of 
#airline_data as well as the percentages (using `.mean()` instead of `.sum()`).
#Order them (increasing or decreasing) using sort_values
#Call `pd.concat` to present these in a single table (DataFrame) with the helpful column names 'count' and '%'
missing = pd.concat([airline_data.isnull().sum(), 100 * airline_data.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by='count', ascending = False)

count         %
Arrival Delay                             393  0.302587
ID                                          0  0.000000
Gate Location                               0  0.000000
Baggage Handling                            0  0.000000
In-flight Entertainment                     0  0.000000
In-flight Wifi Service                      0  0.000000
In-flight Service                           0  0.000000
Food and Drink                              0  0.000000
Cleanliness                                 0  0.000000
Leg Room Service                            0  0.000000
Seat Comfort                                0  0.000000
On-board Service                            0  0.000000
Online Boarding                             0  0.000000
Gender                                      0  0.000000
Check-in Service                            0  0.000000
Ease of Online Booking                      0  0.000000
Departure and Arrival Time Convenience      0  0.000000
Departure Delay                             0  0.000000
Flight Distance                             0  0.000000
Class                                       0  0.000000
Type of Travel                              0  0.000000
Customer Type                               0  0.000000
Age                                         0  0.000000
Satisfaction                                0  0.000000

In [11]:
# I want to confirm the number of unique data we have per ID.
# I will use the `nunique` method

airline_data[['ID']].nunique()

ID    129880
dtype: int64

This shows us that we have 129,880 unique customers who were surveyed.

Additionally, although it appears that most of our data does not have any missing values, we can see that we do have 393 missing values in Arrival Delay.  That is something I will want to address.

I also want to look at the columns of data, and see whether we have any values within each that will need to be renamed.  To do so, I will first determine what each column is, and then I will look at the unique values within each column.

In [12]:
airline_data.columns

Index(['ID', 'Gender', 'Age', 'Customer Type', 'Type of Travel', 'Class',
       'Flight Distance', 'Departure Delay', 'Arrival Delay',
       'Departure and Arrival Time Convenience', 'Ease of Online Booking',
       'Check-in Service', 'Online Boarding', 'Gate Location',
       'On-board Service', 'Seat Comfort', 'Leg Room Service', 'Cleanliness',
       'Food and Drink', 'In-flight Service', 'In-flight Wifi Service',
       'In-flight Entertainment', 'Baggage Handling', 'Satisfaction'],
      dtype='object')

In [13]:
airline_data["ID"].unique()

array([     1,      2,      3, ..., 129878, 129879, 129880])

In [14]:
airline_data["Gender"].unique()

array(['Male', 'Female'], dtype=object)

In [15]:
airline_data["Age"].unique()

array([48, 35, 41, 50, 49, 43, 60, 38, 28, 27, 24,  9, 52, 70, 61, 11, 42,
       14, 56, 77, 51, 47, 58, 40, 21, 33, 39, 54, 46, 45, 23, 20, 44, 37,
       36, 57, 31, 53, 25, 55, 72, 19, 65, 32, 80,  7, 34, 18, 12, 26, 17,
       66, 68, 67, 59, 69, 16, 22, 64, 29, 10, 71, 30, 62, 63,  8, 15, 73,
       78, 74, 79, 13, 76, 75, 85])

In [16]:
airline_data["Customer Type"].unique()

array(['First-time', 'Returning'], dtype=object)

In [17]:
airline_data["Type of Travel"].unique()

array(['Business', 'Personal'], dtype=object)

In [18]:
airline_data["Class"].unique()

array(['Business', 'Economy', 'Economy Plus'], dtype=object)

In [19]:
airline_data["Departure and Arrival Time Convenience"].unique()

array([3, 2, 4, 1, 5, 0])

In [20]:
airline_data["Ease of Online Booking"].unique()

array([3, 2, 4, 1, 5, 0])

In [21]:
airline_data["Check-in Service"].unique()

array([4, 3, 1, 2, 5, 0])

In [22]:
airline_data['Online Boarding'].unique()

array([3, 5, 4, 2, 1, 0])

In [23]:
airline_data['Gate Location'].unique()

array([3, 2, 4, 1, 5, 0])

In [24]:
airline_data['On-board Service'].unique()

array([3, 5, 4, 2, 1, 0])

In [25]:
airline_data['Seat Comfort'].unique()

array([5, 4, 3, 2, 1, 0])

In [26]:
airline_data['Leg Room Service'].unique()

array([2, 5, 3, 4, 1, 0])

In [27]:
airline_data['Cleanliness'].unique()

array([5, 4, 3, 2, 1, 0])

In [28]:
airline_data['Food and Drink'].unique()

array([5, 3, 4, 2, 1, 0])

In [29]:
airline_data['In-flight Service'].unique()

array([5, 3, 4, 1, 2, 0])

In [30]:
airline_data['In-flight Wifi Service'].unique()

array([3, 2, 4, 1, 0, 5])

In [31]:
airline_data['In-flight Entertainment'].unique()

array([5, 3, 4, 2, 1, 0])

In [32]:
airline_data['Baggage Handling'].unique()

array([5, 3, 4, 1, 2])

In [33]:
airline_data['Satisfaction'].unique()

array(['Neutral or Dissatisfied', 'Satisfied'], dtype=object)

Wonderful!  It appears that my data is fairly clean.  The values are formatted correctly (for example, they are all either between 0 and 5, or Females/Males instead of F or M, mislabeled, or misspelled.  As a result, I will not need to worry about renaming this data.

In [35]:
airline_data.shape

(129880, 24)

Something I want to look at is the Satisfaction column:

In [36]:
print(airline_data['Satisfaction'].value_counts())

Neutral or Dissatisfied    73452
Satisfied                  56428
Name: Satisfaction, dtype: int64


So according to the above, it appears that there is a fairly even split between the number of satisfied and neutral or dissatisfied customers.

I think I want to split up the data a bit to make it easier to look at, so I'll split it between the categorical data and the numerical data.

In [38]:
numerical_cols = airline_data.loc[:, 'Departure and Arrival Time Convenience': 'Baggage Handling'].columns
print(numerical_cols)

Index(['Departure and Arrival Time Convenience', 'Ease of Online Booking',
       'Check-in Service', 'Online Boarding', 'Gate Location',
       'On-board Service', 'Seat Comfort', 'Leg Room Service', 'Cleanliness',
       'Food and Drink', 'In-flight Service', 'In-flight Wifi Service',
       'In-flight Entertainment', 'Baggage Handling'],
      dtype='object')


In [39]:
airline_data[numerical_cols].describe().T

count      mean       std  min  \
Departure and Arrival Time Convenience  129880.0  3.057599  1.526741  0.0   
Ease of Online Booking                  129880.0  2.756876  1.401740  0.0   
Check-in Service                        129880.0  3.306267  1.266185  0.0   
Online Boarding                         129880.0  3.252633  1.350719  0.0   
Gate Location                           129880.0  2.976925  1.278520  0.0   
On-board Service                        129880.0  3.383023  1.287099  0.0   
Seat Comfort                            129880.0  3.441361  1.319289  0.0   
Leg Room Service                        129880.0  3.350878  1.316252  0.0   
Cleanliness                             129880.0  3.286326  1.313682  0.0   
Food and Drink                          129880.0  3.204774  1.329933  0.0   
In-flight Service                       129880.0  3.642193  1.176669  0.0   
In-flight Wifi Service                  129880.0  2.728696  1.329340  0.0   
In-flight Entertainment                 129880.0  3.358077  1.334049  0.0   
Baggage Handling                        129880.0  3.632114  1.180025  1.0   

                                        25%  50%  75%  max  
Departure and Arrival Time Convenience  2.0  3.0  4.0  5.0  
Ease of Online Booking                  2.0  3.0  4.0  5.0  
Check-in Service                        3.0  3.0  4.0  5.0  
Online Boarding                         2.0  3.0  4.0  5.0  
Gate Location                           2.0  3.0  4.0  5.0  
On-board Service                        2.0  4.0  4.0  5.0  
Seat Comfort                            2.0  4.0  5.0  5.0  
Leg Room Service                        2.0  4.0  4.0  5.0  
Cleanliness                             2.0  3.0  4.0  5.0  
Food and Drink                          2.0  3.0  4.0  5.0  
In-flight Service                       3.0  4.0  5.0  5.0  
In-flight Wifi Service                  2.0  3.0  4.0  5.0  
In-flight Entertainment                 2.0  4.0  4.0  5.0  
Baggage Handling                        3.0  4.0  5.0  5.0

I also want to look at the other numerical data (not just the data the customers marked between 0m,and 5).  I'll call this data flight_cols:

In [40]:
flight_cols = ['Age', 'Flight Distance', 'Departure Delay', 'Arrival Delay']

In [41]:
airline_data[flight_cols].describe().T

count         mean         std   min    25%    50%  \
Age              129880.0    39.427957   15.119360   7.0   27.0   40.0   
Flight Distance  129880.0  1190.316392  997.452477  31.0  414.0  844.0   
Departure Delay  129880.0    14.713713   38.071126   0.0    0.0    0.0   
Arrival Delay    129487.0    15.091129   38.465650   0.0    0.0    0.0   

                    75%     max  
Age                51.0    85.0  
Flight Distance  1744.0  4983.0  
Departure Delay    12.0  1592.0  
Arrival Delay      13.0  1584.0

I know from earlier that Arrival Delay has 393 missing observations.

Since the median delay is 0 min, it's safe to fill in the missing values with 0.

In [42]:
airline_data['Arrival Delay'].fillna(0, inplace = True)
print(airline_data['Arrival Delay'].isnull().sum())

0


Now I want to convert Satisfaction to numeric to make it easier to work with moving forward:

In [43]:
airline_data['Satisfaction'] = airline_data['Satisfaction'].apply(lambda x: 1 if (x == 'satisfied') else 0)

## 3. Export Cleaned Data

In [44]:
airline_data.to_csv('/Users/lauren/Desktop/airline_data_cleaned.csv')

Now that I have explored and wrangled my data, it is now ready for me to move on to the next step...exploratory data analysis!